flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/SubtaskGatewayImpl.java

```diff
@@ -64,6 +69,17 @@
                 subtaskAccess.createEventSendAction(serializedEvent);
 
         final CompletableFuture<Acknowledge> result = new CompletableFuture<>();
+        result.whenCompleteAsync(
```

Comment:
What if this is a dated gateway ? Will the result be counted for next checkpoint ?

Reply:
Which checkpoint the events are attributed to, and which gateway they target is independent, that is a nice part in the design.

Here we only need to care about the checkpoint, which means that we need to obtain these pending futures in the same atomic operation as we close the event valve (which is analogous to the point of injecting the checkpoint barrier).
All events before the valve shuts are tracked, no events after the valve are tracked.

The event sending action is enqueued in the main thread executor, same as the valve-closing-action, so the events that are sent later cannot overtake that action.

What gateway the events target is not important here. They may very well target an outdated gateway, in which case they will fail immediately (as soon as the `Execution` is in a terminal state, or the RPC Ask-Timeout fires.

Reply:
I guess I see the correctness in my brain modeling.
* `SubtaskGateway` is eagerly binding to `Execution` so that there will be no mis-targeting events. 
* Both checkpoint-completion and event sending are sending ordered to main thread for execution. So, rendezvous in checkpointing from user side is preserved in holder side. This guarantees that "the events that are sent later cannot overtake that action".
* `createEventSendAction` constructs a lazy sending action to execute in main thread. This sending action will inspect execution state(`Execution.state` or `IncompleteFuturesTracker.failureCause`) to **fail immediately to not become pending**.

It might be nice(eg. correctness evaluation) in design part, but I would say it is pretty un-straightforward in implementation, at least for me 😅 . Here is what I considered as un-straightforward:
* `Execution.state` inspection in sending action is pretty important here. As it contribute to fail immediately. Normally, it should not. That say, a delayed failure should not contribute to correctness. I could not even know whether `taskExecution.getState()` is a purposed duplication for similar inspection in `sendOperatorEvent`.
* Base on above(eg. delayed failure), `taskExecution.getTerminalStateFuture().thenAccept()` becomes crucial. It is not a safeguard to "speed up things", but a safeguard to correctness now. Without this, a delayed failure will contribute to next checkpoint.

I think the key part is that `OperatorEventValve` is neutral to(eg. has no knowledge of) `Execution`, so some work has to be done outside it to avoid dated sending action contributing to incomplete futures which are inspected in next checkpoint. I have no strong objection to go current approach, but I would suggest to adjust docs a bit in `ExecutionSubtaskAccess` and/or other places to emphasize that this immediately failure is crucial  to correctness.

Reply:
It takes me some time to conclude this, I hope I was not wrong at somewhere(forgive me then please 😅 ). It is just hard to evaluate all things with multiple futures. But thanks to all existing works, almost every things are single threaded now, it is much better than asynchronous counterparts.

Reply:
To be precise, the "correctness" above should be accuracy(or similar words) of pending future tracking, it does not contribute to exactly-once guarantee of coordinator holder.

I constructed a test case with following changes:
* Change `FutureUtils.forward(sendResult, result)` in `OperatorEventValve.callSendAction` to a delayed forward.
* Fail subtask but never cancel event sending with outdated gateway.

Then I see more checkpoint abortion than usual.